docker/server/Dockerfile.ubuntu

```diff
@@ -40,82 +45,78 @@ ARG DIRECT_DOWNLOAD_URLS=""
-            || exit 1 \
-        ; done \
-        && dpkg -i /tmp/clickhouse_debs/*.deb \
-        && rm -rf /tmp/* ; \
-    fi
-
-# install from a single binary
-RUN if [ -n "${single_binary_location_url}" ]; then \
-        echo "installing from single binary url: ${single_binary_location_url}" \
-        && rm -rf /tmp/clickhouse_binary \
-        && mkdir -p /tmp/clickhouse_binary \
-        && wget --progress=bar:force:noscroll "${single_binary_location_url}" -O /tmp/clickhouse_binary/clickhouse \
-        && chmod +x /tmp/clickhouse_binary/clickhouse \
-        && /tmp/clickhouse_binary/clickhouse install --user "clickhouse" --group "clickhouse" \
-        && rm -rf /tmp/* ; \
-    fi
-
-# A fallback to installation from ClickHouse repository
-RUN if ! clickhouse local -q "SELECT ''" > /dev/null 2>&1; then \
-        apt-get update \
-        && apt-get install --yes --no-install-recommends \
+            || exit 1; \
+        done; \
+        dpkg -i /tmp/clickhouse_/*.deb; \
+    fi; \
+    \
+    # install from a single binary
+    if [ -n "${single_binary_location_url}" ]; then \
+        echo "installing from single binary url: ${single_binary_location_url}"; \
+        wget --progress=bar:force:noscroll "${single_binary_location_url}" -O /tmp/clickhouse/clickhouse; \
+        chmod +x /tmp/clickhouse/clickhouse; \
+        /tmp/clickhouse/clickhouse install --user "clickhouse" --group "clickhouse"; \
+    fi; \
+    \
+    # A fallback to installation from ClickHouse repository
+    if ! clickhouse local -q "SELECT ''" > /dev/null 2>&1; then \
+        apt-get update; \
+        apt-get install --yes --no-install-recommends \
             apt-transport-https \
             ca-certificates \
             dirmngr \
             gnupg2 \
-        && mkdir -p /etc/apt/sources.list.d \
-        && GNUPGHOME=$(mktemp -d) \
-        && GNUPGHOME="$GNUPGHOME" gpg --no-default-keyring \
+            ;\
+        mkdir -p /etc/apt/sources.list.d; \
+        GNUPGHOME=$(mktemp -d); \
+        GNUPGHOME="$GNUPGHOME" gpg --no-default-keyring \
             --keyring /usr/share/keyrings/clickhouse-keyring.gpg \
-            --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 8919F6BD2B48D754 \
-        && rm -rf "$GNUPGHOME" \
-        && chmod +r /usr/share/keyrings/clickhouse-keyring.gpg \
-        && echo "${REPOSITORY}" > /etc/apt/sources.list.d/clickhouse.list \
-        && echo "installing from repository: ${REPOSITORY}" \
-        && apt-get update \
-        && for package in ${PACKAGES}; do \
-            packages="${packages} ${package}=${VERSION}" \
-        ; done \
-        && apt-get install --allow-unauthenticated --yes --no-install-recommends ${packages} || exit 1 \
-        && rm -rf \
+            --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 8919F6BD2B48D754; \
+        rm -rf "$GNUPGHOME"; \
```

Comment:
How is it better than the logical and operator?

Reply:
The `set -ex` makesit equivalent to an and operator but easier to read. It fails on the first error.